src/engine/contactsengine.cpp

```diff
@@ -737,8 +769,21 @@ class JobContactReader : public ContactReader
 
 void JobThread::run()
 {
-    ContactsDatabase database;
-    if (!database.open(QString(QLatin1String("qtcontacts-sqlite-job-%1")).arg(m_databaseUuid), m_nonprivileged, true)) {
+    QString dbId(QStringLiteral("qtcontacts-sqlite%1-job-%2"));
+    dbId = dbId.arg(m_autoTest ? QStringLiteral("-test") : QString()).arg(m_databaseUuid);
+
+    QMutexLocker locker(&m_mutex);
+
+    m_database.open(dbId, m_nonprivileged, m_autoTest);
+    m_nonprivileged = m_database.nonprivileged();
+    m_running = true;
+
+    {
+        MutexUnlocker unlocker(locker);
+        m_wait.wakeOne();
+    }
```

Comment:
What is the unlocking supposed to do here?
I'm asking because whatever it is, I don't think it does it. There's a valid code path where it simply unlocks, wakes the thread, then locks again before the thread has acquired the lock.

I think it's worth creating a separate QWaitCondition like m_waitRunning for this use, so that it's clearly separated from the pendingJobs related wait condition and doesn't depend on timing.

Alternately, could you open the db in the main thread and then pass it to the job thread? We're using an sqlite version that supports using a connection from multiple threads (3.8.5, compiled with THREADSAFE=1), see http://www.sqlite.org/threadsafe.html